### ISSUE TITLE
Ability to merge geopackage files

### ISSUE THREAD

**Issue by mamayoleksandr** [radiantbluetechnologies/rialto-geopackage#50]
Is there any way to merged or append the  geopackage files?

**Comment by mpgerlek** [comment]
I built the underlying framework with that in mind, but I've not gotten around to implementing the public interfaces.

We need to add functions like addTable, deleteTable, and renameTable to the Manager class and then add those features to the command line tool. It's probably ~2 days of work.

**Comment by mpgerlek** [comment]
(see also #3, #7)

**Comment by mamayoleksandr** [comment]
Michael,
thanks for the response. Does that mean that the functionality you're talking about will just add new tile table to gpkg and update metadata, the tile tables themselves won't be merged? 
This all relates to serving of bigger data (TB).

**Comment by mamayoleksandr** [comment]
To clarify question, will it be possible to use big gpkg with single url (table)?

**Comment by mpgerlek** [comment]
Yes, I have been thinking of juststoring multiple, independent 'files' in the DB. I don't have a use case in mind, it just seems like a good thing to offer at some point. 

But it sounds like your are wanting to merge two files into one set of tiles. That's doable, but would be more work -- I'd need to think about for a bit.  

Is this work you'd be interested in attempting? (and what is your use case for Rialto, I'd live to hear what features you need / don't need...)

**Comment by mpgerlek** [comment]
In theory, a gpkg file can be tens of GBs in size, all storing just one really big set of tiles.

**Comment by mamayoleksandr** [comment]
Am looking for open source solution for rendering point cloud in web. The Rialto module for cesium is really nice and almost what we need. The only think to sort is serving big data, eg seamless navigation through big point cloud.
I was thinking on two possible approaches:
1. merged different tile sets into single one, and use it as it is.
2. implement kinda proxy dispetcher which would query according gpkg depending on request. but this also would require changing client side code, and less viable I think.

am gonna try to work on it bit, and will let you know  if have some working stuff.

**Comment by mpgerlek** [comment]
Let me see if I understand --

You have many point clouds right now (say, 100 LAS files) and you want them to combine into one large geopackage table.

**Comment by mamayoleksandr** [comment]
yes. currently for smaller peices (0.5 - 2GB) I can use lasmerge tool to get single las file, than rialto_translate to get gpkg.

Basically think its possible insert tiles from another gpkg, just need to define tile_column and tile_row, and update BBOX, right?

**Comment by mpgerlek** [comment]
I'm assuming you can't use lasmerge for all the input files because it would be just too big (and would be similar problem if we allowed the Rialto cmd line tool to accept 1000 imput LAS files instead of just one). 

You would need to maintain the properties of the gpkg tile tree as you add new tiles to it. Two issues here:

- you need to generate the reduced resolution tiles for the tiles you are adding, and also update the affected parent tiles that are already there. 

- you need to update the patent tile's "child mask" for each new tile you add

I dont think this would be too bad actually. A bunch of SQL adds, and some careful SQL updates. Hmm......

If you wanted to do the implementation, I could give you the pseudo code for what needs to be done. Probably a couple days of implemention work at most.

**Comment by mamayoleksandr** [comment]
Yep, pseudo code would be good. 
Though I didn't write anything on c/c++, only had to work on java, python and all web-around techs, so frankly speaking, there is no guarantee I will implement it at all, but at least I can try.

thanks.

**Comment by mpgerlek** [comment]
Okay, I'll write something up in the morning.

**Comment by mamayoleksandr** [comment]
thank you much.

**Comment by mpgerlek** [comment]
See my notes here: https://github.com/radiantbluetechnologies/rialto-geopackage/wiki/Notes-on-Updating-a-GeoPackage

This is probably more complicated than you want to deal with, sorry. But if you get a start into it, I'm willing to help out.

**Comment by mamayoleksandr** [comment]
Michael,

thanks for the notes, quite helpful. I will try to have a look, however, due to other workload am currently have to postpone working in this area. Gonna pick it up bit later.

thanks,

Alex.

**Comment by mamayoleksandr** [comment]
Hi Michael,

OK, quick update on it: my colleague, a c++ developer briefly took a look at the code, and noticed that data is loaded mainly in memory, which would cause issues on  very large (TB) data sets. Seems like the changes needed to be done are significant and currently we unable to proceed on this task.

thanks,

Alex.

**Comment by mpgerlek** [comment]
Out of curiosity, which parts did you find as memory-limited? (sqlite itself should be fine, but we do know that PDAL in its current form is limited for a number of workflows)

**Comment by hobu** [comment]
PDAL does have the ability to do "one" at-a-time callback reads if you don't want to maintain the entire buffer of a file in memory. This only works on readers though, not entire pipelines, and if the thing you are reading isn't stream-like, you still will pay some penalty (BPF, some db drivers).

https://github.com/PDAL/PDAL/blob/master/test/unit/io/las/LasReaderTest.cpp#L375

**Comment by mamayoleksandr** [comment]
Hi Michael,

I personally didn't look into the code, sorry. Just repeated what said my colleague, so he was not going to proceed. am not good in it.
